### DISCUSSION TITLE
Integration with Visual Studio 2022

### DISCUSSION THREAD

**Discussion by iainwest** [containers/podman#17131] [Q&A]
Has anybody managed to get PodMan working with Visual Studio 2022?  I have attempted and got warnings, after I changed some variables e.g. DOCKER_HOST to podman under wsl2 have now been presented with an Error response from daemon No such file or directory, OCI runtime attempted to invoke a command that was not found.

Just wondering if there are other variables I need to set?

**Comment by rhatdan** [comment]
If you run Podman directly does it work?

If you run docker with DOCKER_HOST set does it work?

**Reply by iainwest** [reply]
Running Podman directly, I am able to launch a container, I don't have docker desktop installed as I removed this before installing Podman.

After a little bit more trial and error, I reset WSL2 and performed the following steps, all from windows terminal

1. `podman stop`
2. `podman machine rm podman-machine-default`
3. `podman machine init`
4. `podman start`

The only other change since I started this was I am now running Visual Studio 2022 Version 17.6.0 Preview 1.0 with .Net 8  Preview 1 installed and my test container is targeting .Net 8.

I can only assume that my first attempt caused some corruption within the setup.

**Comment by VeselovAndrey** [comment]
Same issue for me. Win11 + VS 2022 latest + Podman Desktop. Also have a VS Code. All tooling can't see Podman (but it's running) and asking to install Docker.  Are there any manual how to configure tooling?

**Reply by cinorid** [reply]
@iainwest @rhatdan @VeselovAndrey
it's because the visual studio is trying to find docker.exe. the easiest way to fix this is to make a copy of podman.exe and name it docker.exe.

**Reply by rhatdan** [reply]
Could someone open an issue with visual studio to look for podman.exe as well?

**Reply by VeselovAndrey** [reply]
@Cinorid Didn't work for me. Only working way is to get "docker.exe" from Docker CE and add it location to the PATH env var.

**Reply by thawkins** [reply]
How do we get VS2022 to interact with a podman installation hosted inside a wsl2 distribution, ie the podman commands are executed from the wsl2 distro. 

This works just fine in VSC, and any Jetbrains tooling, but does not work on VS2022. VS seems to be determined to find docker/podman in the windows host and not in the designated WSL distribution. We are using devcontainers hosted with podman inside our custom wsl2 distribution.

**Comment by jeroen-mostert** [comment]
For others trying to make VS play nice with Podman, aside from copying `podman.exe` a little more work is needed to get it to work with `podman-compose` (much like Docker, it's hardcoded to look for `docker-compose` and moreover makes some assumptions that don't work for `podman-compose`). I've cobbled together a little wrapper for `docker-compose` at https://github.com/jeroen-mostert/vs-podman-wrapper , which also has instructions for getting the rest to work. A cleaner integration is certainly possible, but writing VS extensions is not in any way fun.

**Comment by gudge25** [comment]
any  update?

**Reply by rhatdan** [reply]
Has anyone approached Visual Stodio to get them to work with Podman?

**Reply by gudge25** [reply]
so  the solution is 
1- Find out  DOCKER_HOST 
https://podman-desktop.io/docs/migrating-from-docker/using-the-docker_host-environment-variable
2- Downgrade  Docker  Extension  to v1.22.2
3- 'Docker Host '  in 'Docker  Extension' settings    to      unix://..../podman.sock (from step 1)
4- set 'Docker Path' in 'Docker  Extension' to  podman executive program   example  /usr/local/bin/podman

restart VS Code

**Reply by rhatdan** [reply]
I mean natively support Podman not in Docker compatibility mode.

**Comment by soerennielsen** [comment]
Also wanting to run podman with VS 2022.

It is however sadly a common problem in these threads that some people Tak about VS code and others VS 2022.

Two very different IDEs that share nothing in terms of tooling setup.

Thanks to original poster. Will try some of the steps here. For me the acceptance criteria is that I can start from VS 2022 and execute, debug (auto attach remote debugger) and we'll.. Build/package.

**Comment by pozy** [comment]
If anyone is interested, here's related feature request for Visual Studio ([link](https://developercommunity.visualstudio.com/t/visual-studio-container-tools-depend-on-docker-des/1537846)). More votes - bigger chance VS team would implement it.

**Comment by GarretWayne** [comment]
Took a while and asked around, however the current working strategy for us is looking like this: 
- Installed packages:
   - `Docker.CLI` 
      - to communicate with Podman Daemon (that is in WSL)
   - `Podman` (can be Desktop or just "`podman-machine`" + "`podman-cli`")
     - desktop fully installed with WSL fully working
     - and a configured podman-machine

The idea behind the whole thing is Podman's API forwarding capabilities (see [this section](https://github.com/containers/podman/blob/main/docs/tutorials/podman-for-windows.md#using-api-forwarding)) 
and the idea behind switching out the thin client that is used on windows to communicate with Podman's Daemon 

Also if your "`docker info`" say something along the line as "IPV4 port-forwarding is disabled", you can use `"sudo sysctl net.ipv4.ip_forward=1"` to enable it. (Although a fair warning as all new podman-machines might require setting this manually at first launch).

This setup resulted in a fully working Visual Studio Container tab, and the Terminal to respond to all docker commands, whilst all containers are hosted in WSL by Podman.

**Reply by soerennielsen** [reply]
Would love a write up on how to set this up

**Reply by GarretWayne** [reply]
I will be back with a guide later ;)

**Reply by soerennielsen** [reply]
;) looking forward to it. 

I got it somewhat working: 

- Installed podman (followed their guide)
- symlinked docker.exe to podman.exe (windows) to make visual studio builds go the right place. 

It works in visual studio (not code!), build + debugging. Container tools do not work and also it wants to start docker desktop initially. 
Requires podman to be running (which will block docker desktop from running).
Clunky but working. There will likely be more details not working.

**Reply by tlunsfordCXP** [reply]
Any progress on this in recent months? I'm about to go down this rabbit hole and any new info would be helpful.

**Reply by GarretWayne** [reply]
## Disclaimer: 
- The short version should be the first approach to start with, as the long-one I did with a podman desktop first approach. However, the long approach is still heavily riddled with powershell commands.

--------------------------------------------------------------------------------------------------
## Short version (package first approach in Powershell)
- For an examplanation on what each command does, please refer to the "long version"
1. Prerequisites (somewhere with a restart when the command line says so):
     1.   `dism.exe /online /enable-feature /featurename:VirtualMachinePlatform /all /norestart`
     2. `dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart`
     3. `wsl --install`
     4. `wsl --update`
2. `winget.exe install --id Docker.DockerCLI`
3. `winget.exe install --id RedHat.Podman `
4. `winget.exe install --id RedHat.Podman-Desktop` (optional, but nice to have UI)
5. `podman machine init`
6. `wsl sudo sysctl net.ipv4.ip_forward=1`

### Testing
- `docker run --rm -d -p 8080:80 --name httpd docker.io/library/httpd`
- open `localhost:8080` in your browser

--------------------------------------------------------------------------------------------------
# Docker CLI with Podman Daemon
a.k.a  UI First Approach (or rather the "problems first" approach)

## Steps Guide
This guide is based on a freshly reset Win 11 install, that is akin to a new installation too. Thus, it is based on the exact steps I made, based on the problems I have encountered to have a fully functioning setup.
## Step 1. Download Podman Desktop
### Sources:
- [Podman website](https://podman-desktop.io/blog/podman-desktop-release-1.7#:~:text=Podman%20Desktop%201.7%20is%20now%20available.%20Click%20here%20to%20download%20it!) 
- [Podman WingeUi Package](https://marticliment.com/wingetui/share?pid=RedHat.Podman-Desktop&pname=Podman%20Desktop&psource=Winget%3A%20winget) 

## Step 2. Install Podman Desktop
- 2.1 Run the normal installation until you get to the landing page such as below:
    ![image](https://github.com/user-attachments/assets/d1ca851c-a09a-4cc3-9d15-cd00e3d36444)

## Step 3. Configure Podman Client and WSL features
- 3.1 Press "Go to Podman Desktop"
- 3.2 on the next page you should see:
    ![image](https://github.com/user-attachments/assets/8f916cb8-7cb0-426c-83db-0fe11e59b9dd)
- 3.3 On this page you should press the "Install" button
   - 3.3.1 the following error might occur:
       ![image](https://github.com/user-attachments/assets/80e67a28-8b09-4220-9079-8caccace1389)
       - this error means that the Virtual Machine Platform feature has not yet been enabled
       - Solution: https://learn.microsoft.com/en-us/windows/wsl/install-manual#step-3---enable-virtual-machine-feature 
       - Run the following command in Powershell: `dism.exe /online /enable-feature /featurename:VirtualMachinePlatform /all /norestart`
- 3.4 After enabling the VM Platform Feature, you might encounter this error:
    ![image](https://github.com/user-attachments/assets/7793f443-4eb7-4101-871f-66abf802817e)
    - 3.4.1 This problem means that there are no installed "Windows Subsystem for Linux" versions installed
       - furthermore, this error might be paired with these messages (during the regular setup process //see point 3.2 and choosing "Setup")
           ![image](https://github.com/user-attachments/assets/3915482c-c218-4e8c-bf11-e84374a1e32e)
           ![image](https://github.com/user-attachments/assets/be727d22-be15-49da-a538-4a306f71d407)
     - 3.4.2 At this point, you might also receive an error from Powershell for wsl --install command:
     
           > -Class not registered
           Error code: Wsl/CallMsi/REGDB_E_CLASSNOTREG

           - this means that you don't have the WSL feature enable at all
                - **Solution**:
                    - Run this command: `dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart`
                    - 
- 3.5 Now you can use the `wsl --install` command for Powershell
     - 3.5.1 also, don't forget to use wsl --update thereafter, to make sure you have the latest updates for WSL
- 3.6 Check the requirements again, if all goes well you should be greeted with:
        ![image](https://github.com/user-attachments/assets/258361e9-43ca-4231-acbb-0ad98386f12b)


    - This means, that the Podman Client is not installed
    - Proceed with the installation (press "Yes"), you'll be greeted with:
       ![image](https://github.com/user-attachments/assets/e251c79e-55eb-4446-af0d-ac1f3ec283dc)
    - Proceed with the defaults and press "Install"
## Step 4 Setup Podman Machine
- Podman machine is the daemon that is running in WSL, that we can communicate with in order to launch containers
- 4.1 Inside Podman Desktop GUI, Navigate to Settings(Cog-icon, button left)/Resources
   - 4.1.1 As you don't have any podman machines yet you'll be greeted with:
![image](https://github.com/user-attachments/assets/e45daeda-1d81-4098-8542-e9c16e9c5911)
![image](https://github.com/user-attachments/assets/c8f9162c-174f-4221-a2e0-06e0d78d02ce)
- Proceed with the defaults
   ![image](https://github.com/user-attachments/assets/a9b00089-b8d1-4c0f-ac5b-3fb9a341aee8)

- 4.2 at this point you might encounter:
   ![image](https://github.com/user-attachments/assets/c10e14d1-c953-4c2d-83ac-026fa2963221)
   - For me, this was caused because the installer didn't actually manage to install WSL's optional components
        - Debug steps included trying to init from GUI, which produced the same error.
   - **Solution**:
        - In PowerShell, running the command "`podman machine init`"
            - windows prompts for:
                ![image](https://github.com/user-attachments/assets/0efa3cb3-c53a-43c4-a773-6a609f66cbde)
        - proceeding with the install will produce the following:
             ![image](https://github.com/user-attachments/assets/9e63829f-ac08-4f95-9841-34359a5beef7)
        - At this point you should restart your machine
- 4.3 at this point you should be able to:
    - start a podman machine successfully with Powershell command: `podman machine init` (inits default machine)
    - start / stop the podman machine with Powershell command:` podman machine start` / `podman machine stop`
## Step 5 Install Docker.Cli
- 5.1 Install Package: [https://marticliment.com/wingetui/share?pid=docker-cli&pname=Docker Cli&psource=Chocolatey](https://marticliment.com/wingetui/share?pid=docker-cli&pname=Docker%20Cli&psource=Chocolatey) 

- 5.2 Make sure that Powershell answers correctly for command: `docker info`

- 5.3 If at the end of docker info you see:
    ![image](https://github.com/user-attachments/assets/60a76646-1301-4b78-9841-c0542f0183aa)

- This means that you have to open enable IPv4 port forwarding
- **Solution**:
    - Open Powershell and use command: `wsl `(opens the installed WSL system's console)
    - Use command: `sudo sysctl net.ipv4.ip_forward=1` (enbles IPv4 port forwarding)
    - You can exit the WSL console by typing `exit `(might need to type it twice to get back to Powershell)
    - Make sure `docker info` doesn't have the aforementioned error anymore
## DONE
- at this point you should be able to pull images in Powershell, using the `docker `keyword
    - example:
        - by running `docker run --rm -d -p 8080:80 --name httpd docker.io/library/httpd`
            - you should be able to see the running container in Podman Desktop's GUI
            - as well as get the text response "Ok" when connecting to `localhost:8080`, confirming that the container is running
 - furthermore, in Visual Studio IDE, you should have a fully functioning "View/Containers" tab
    - if you have tried to run a solution that is supposed to have a container publish before
    - you might encounter a popup saying "docker.exe" cannot be found
    - however, after closing that window the solution should run as intended
    - this problem was no longer happening on a freshly installed VS IDE, and a freshly cloned Solution, so it can be possible a simple deletion of VS specific files in the solution folder, might solve the popup (otherwise it's a WIP to find out where VS actually gets the idea from, wrongly)

**Reply by GarretWayne** [reply]
@tlunsfordCXP

I dropped my guide that I did a while back. I hope this can help the rabbit hole problem ;)

It has been stable for me since then and this is my current setup.

**Reply by Ungerfall** [reply]
I followed the guide, but on `docker run --rm -d -p 8080:80 --name httpd docker.io/library/httpd` getting 
>docker: no matching manifest for windows/amd64 10.0.22631 in the manifest list entries.

Was it because I didn't remove Docker Desktop correctly? Or should I follow @soerennielsen instruction to create a symb link?

**Comment by jrlost** [comment]
Are/were you running windows containers?

**Reply by soerennielsen** [reply]
I didn't.

**Reply by Ungerfall** [reply]
After I installed the docker cli via winget The `docker info` gave
```yaml
OSType: windows
```
Then I tried to find how to change the OSType, and it seemed impossible without Docker Desktop.
Now I checked after the restart
```yaml
OSType: linux
```

**Reply by soerennielsen** [reply]
? but the OSType depends on the docker/podman host. 
For both I would use WSL which in turn mean that it is linux containers. 

You are gonna hit compatibilty issues when you have both installed. You _can_ but make sure that one is only used at any one time as they will fight over the named pipes otherwise. 

It has been a few years since I dabled in windows containers, but it is using hyper-v, not wsl to do that. Worked okish with docker, never trided with podman - I suspect it is not supported.
